packages/aws-rfdk/lib/core/lambdas/nodejs/version-provider/handler.ts

```diff
@@ -7,52 +7,253 @@
+
+  /**
+   * The URI of the Deadline client installer for Windows.
+   */
+  readonly WindowsClientInstaller: string;
+
+  /**
+   * The URI of the Deadline repository installer for Windows.
+   */
+  readonly WindowsRepositoryInstaller: string;
+
+  /**
+   * The URI of the Deadline certificate installer for Windows.
+   * @default - No installer provided
+   */
+  readonly WindowsCertificateInstaller?: string;
+
+  /**
+   * The full version of the Deadline Docker recipes, saved as "major.minor.release.patch". For example: "10.1.10.6".
+   */
+  readonly LinuxDeadlineDockerVersion: string;
+
+  /**
+   * The URI of the Deadline Docker recipes for Linux.
+   */
+  readonly LinuxDockerRecipes: string;
+}
 
+/**
+ * This custom resource will parse and return the S3 URI's of the Deadline installers and Docker recipes for use in
+ * any constructs that need to install Deadline.
+ */
 export class VersionProviderResource extends SimpleCustomResource {
   readonly versionProvider: VersionProvider;
 
-  constructor(indexFilePath?: string) {
+  constructor() {
     super();
-    this.versionProvider = new VersionProvider(indexFilePath);
+    this.versionProvider = new VersionProvider();
   }
 
   /**
    * @inheritdoc
    */
-  /* istanbul ignore next */ // @ts-ignore
   public validateInput(data: object): boolean {
-    return this.versionProvider.implementsIVersionProviderProperties(data);
+    return this.implementsIVersionProviderResourceProperties(data);
   }
 
   /**
    * @inheritdoc
    */
   // @ts-ignore  -- we do not use the physicalId
-  public async doCreate(physicalId: string, resourceProperties: IVersionProviderProperties): Promise<Map<Platform, IVersionedUris>> {
-    return await this.versionProvider.getVersionUris(resourceProperties);
+  public async doCreate(physicalId: string, resourceProperties: IVersionProviderResourceProperties): Promise<FlatVersionedUriOutput> {
+    const deadlinePlatFormVersionedUris = await this.versionProvider.getVersionUris({
+      versionString: resourceProperties.versionString,
+      product: Product.deadline,
+    });
+
+    const deadlineLinux = deadlinePlatFormVersionedUris.get(Platform.linux)!;
+    const deadlineLinuxUris = deadlineLinux.Uris;
+    const deadlineMac = deadlinePlatFormVersionedUris.get(Platform.mac)!;
+    const deadlineMacUris = deadlineMac.Uris;
+    const deadlineWindows = deadlinePlatFormVersionedUris.get(Platform.windows)!;
+    const deadlineWindowsUris = deadlineWindows.Uris;
+
+    const deadlineDockerPlatFormVersionedUris = await this.versionProvider.getVersionUris({
+      versionString: resourceProperties.versionString,
+      product: Product.deadlineDocker,
+      platform: Platform.linux,
+    });
+
+    const deadlineDockerLinux = deadlineDockerPlatFormVersionedUris.get(Platform.linux)!;
+    const deadlineDockerLinuxUris = deadlineDockerLinux.Uris;
+
+    const s3Bucket = this.parseS3BucketName(deadlineLinuxUris.bundle);
+
+    return {
```

Comment:
This doesn't need to return *everything* that is in the metadata file -- it only needs to return the values that would/could be used in an RFDK application. More importantly, we don't want to be breaking the `IVersion` interface -- it's a breaking change, and that would require a major version bump of the RFDK. 

There is a way to do this that won't entail a breaking change.
1. Change the return value of this function to match a flattened `IVersion`. ex:
```
return {
  MajorVersion: ... , // -> IVersion.majorVersion
  MinorVersion: ... , // -> IVersion.minorVersion
  ReleaseVersion: ... , // -> IVersion.releaseVersion
  LinuxPatchVersion: ... , // -> IVersion.linuxInstallers.patchVersion
  LinuxRepositoryS3Bucket: ..., // -> IVersion.linuxInstallers.repository.s3Bucket
  LinuxRepositoryObjectKey: ..., // -> IVersion.linuxInstallers.repository.objectKey
```

We don't currently need the docker bundle information, or the client installer information, or the mac information, or the Windows information. We can always add those later if/when the need arises and we extend the `IVersion` interface to contain that information.

2. Undo all of the breaking changes to `IVersion` and `VersionQuery`.

3. In the implementation of the `VersionQuery` constructor, just invoke this Custom Resource and put the returned values directly into the appropriate spots in the `IVersion`.

Reply:
This is fine, but if we ever need to add the Mac, Windows, or Deadline Docker recipes in the future, we'd likely need to make a breaking change since the patch versions aren't guaranteed to be the same. I think it makes more sense to have a breaking change now that allows this to be extensible to support those in the future, rather than breaking at a later time. 

Maybe most of the other Platform installers and the recipes are unlikely to be needed, so it's fine not to worry about supporting them, but do we see this construct also holding the Docker containers when we begin publishing those? We may want to consider that use case now.

Reply:
Actually I see how my last comment isn't correct. I was thinking `IVersion` extends `IPatchVersion`, but it extends `IReleaseVersion`. We could add additional fields to the `IVersion` interface to support the other use cases without breaking backwards compatibility.

It doesn't make sense for `linuxFullVersionString()` to be a method though, since the full value needs to be returned from the CustomResource rather than combining the version parts into a string. The version parts will be Tokens, so trying to detokenize them in `linuxFullVersionString()` wouldn't work.

Reply:
> we'd likely need to make a breaking change since the patch versions aren't guaranteed to be the same.

Patch versions are allowed to vary, but you caught on to that with the structuring of `IReleaseVersion` and `IPatchVersion` -- a difference in major/minor/release versions between Windows, Linux, and macOS should be treated as an error. Realistically, it'll never happen unless Deadline discontinues support for an operating system.

> It doesn't make sense for linuxFullVersionString() to be a method though, since the full value needs to be returned from the CustomResource rather than combining the version parts into a string. 

AFAIK, this isn't true. CDK lets you combine tokens in a string. For example, this should work:
```ts
const major = Token.asString(deadlineResource.getAtt('MajorVersion'));
const minor = Token.asString(deadlineResource.getAtt('MinorVersion'));
const release = Token.asString(deadlineResource.getAtt('ReleaseVersion'));
const patch = Token.asString(deadlineResource.getAtt('LinuxPatchVersion'));

const linuxVersionString = `${major}.${minor}.${release}.${patch}`;
```